chain/txValidator.go

```diff
@@ -264,107 +297,210 @@ func VerifyTransactionWithLedger(txn *transaction.Transaction) error {
-			if err != nil {
-				return err
-			}
-			if Fixed64(coinbase.Amount) != GetRewardByHeight(header.UnsignedHeader.Height)+donation.Amount {
-				return errors.New("Mining reward incorrectly.")
-			}
+		case pb.TransferAssetType:
+			transfer := payload.(*pb.TransferAsset)
+			amount = Fixed64(transfer.Amount)
 		case pb.RegisterNameType:
 			namePayload := payload.(*pb.RegisterName)
 
 			name := namePayload.Name
-			if _, ok := registeredNames[name]; ok {
-				return errors.New("[VerifyTransactionWithBlock], duplicate name exist in block.")
-			}
-			registeredNames[name] = struct{}{}
+			delete(bvs.registeredNames, name)
 
 			registrant := BytesToHexString(namePayload.Registrant)
-			if _, ok := nameRegistrants[registrant]; ok {
-				return errors.New("[VerifyTransactionWithBlock], duplicate registrant exist in block.")
-			}
-			nameRegistrants[registrant] = struct{}{}
+			delete(bvs.nameRegistrants, registrant)
 		case pb.DeleteNameType:
 			namePayload := payload.(*pb.DeleteName)
 
+			name := namePayload.Name
+			delete(bvs.registeredNames, name)
+
 			registrant := BytesToHexString(namePayload.Registrant)
-			if _, ok := nameRegistrants[registrant]; ok {
-				return errors.New("[VerifyTransactionWithBlock], duplicate registrant exist in block.")
-			}
-			nameRegistrants[registrant] = struct{}{}
+			delete(bvs.nameRegistrants, registrant)
 		case pb.SubscribeType:
 			subscribePayload := payload.(*pb.Subscribe)
 			topic := subscribePayload.Topic
-			key := subscription{topic, SubscriberString(subscribePayload)}
-			if _, ok := subscriptions[key]; ok {
-				return errors.New("[VerifyTransactionWithBlock], duplicate subscription exist in block")
-			}
-			subscriptions[key] = struct{}{}
+			bucket := subscribePayload.Bucket
+			key := subscription{topic, bucket, BytesToHexString(subscribePayload.Subscriber), subscribePayload.Identifier}
+			delete(bvs.subscriptions, key)
 
-			if _, ok := subscriptionCount[topic]; !ok {
-				subscriptionCount[topic] = DefaultLedger.Store.GetSubscribersCount(topic, subscribePayload.Bucket)
-			}
-			if subscriptionCount[topic] >= transaction.SubscriptionsLimit {
-				return errors.New("[VerifyTransactionWithBlock], subscription limit exceeded in block.")
+			bvs.subscriptionCount[topic]--
+
+			if bvs.subscriptionCount[topic] == 0 {
+				delete(bvs.subscriptionCount, topic)
 			}
-			subscriptionCount[topic]++
 		case pb.GenerateIDType:
 			generateIdPayload := payload.(*pb.GenerateID)
+			amount = Fixed64(generateIdPayload.RegistrationFee)
 			publicKey := BytesToHexString(generateIdPayload.PublicKey)
-			if _, ok := generateIDs[publicKey]; ok {
-				return errors.New("[VerifyTransactionWithBlock], duplicate GenerateID txns in block.")
-			}
-			generateIDs[publicKey] = struct{}{}
+			delete(bvs.generateIDs, publicKey)
 		}
 
-		return nil
-	})
-
-}
+		if amount > 0 || fee > 0 {
+			if _, ok := bvs.totalAmount[sender]; ok {
+				bvs.totalAmount[sender] -= amount + fee
```

Comment:
need to check that bvs.totalAmount[sender] > amount + fee?

Reply:
and if bvs.totalAmount[sender] is not existed, how to deal with it?

Reply:
Seems that it's checked in VerifyTransactionWithBlock. This function (CleanSubmittedTransactions) is to rollback BlockValidationState state

Reply:
Seems that bvs.totalAmount[sender] should always exist if txns has been applied to BlockValidationState, so we should probably return error if bvs.totalAmount[sender] does not exist or bvs.totalAmount[sender] < amount + fee. @trueinsider

Reply:
done